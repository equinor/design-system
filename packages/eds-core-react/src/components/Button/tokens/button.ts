import { tokens } from '@equinor/eds-tokens'
import { ButtonToken } from '../Button.types'

const {
  typography: {
    navigation: { button: buttonTypography },
  },
  colors: {
    interactive: {
      disabled__text: { rgba: disabledTextColor },
      disabled__border: { rgba: disabledBorderColor },
      disabled__fill: { rgba: disabledColor },
    },
  },
  spacings: {
    comfortable: { medium },
  },
  shape: {
    button: { minHeight: buttonHeight, borderRadius: buttonBorderRadius },
    _modes: {
      compact: {
        button: { minHeight: compactButtonHeight },
      },
    },
  },
  interactions: { focused: outline },
  clickbounds: {
    default__base: clicboundHeight,
    compact__standard: compactClickboundHeight,
  },
} = tokens

export const button: ButtonToken = {
  background: 'transparent',
  height: `var(--eds_button__height, ${buttonHeight})`,
  typography: {
    ...buttonTypography,
    textAlign: 'center',
    fontSize: `var(--eds_button__font_size, ${buttonTypography.fontSize})`,
  },
  border: {
    type: 'border',
    width: `var(--eds_button__border_width, 1px)`,
    color: 'transparent',
    radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
    style: 'solid',
  },
  spacings: {
    top: 'var(--eds_button__padding_y, 0)',
    bottom: 'var(--eds_button__padding_y, 0)',
    left: `var(--eds_button__padding_x, ${medium})`,
    right: `var(--eds_button__padding_x, ${medium})`,
  },
  clickbound: {
    height: clicboundHeight,
    width: '100%',
    offset: {
      top: `${(parseInt(clicboundHeight) - parseInt(buttonHeight)) / 2 + 1}px`,
      left: '0',
    },
  },
  entities: {
    icon: {
      height: 'var(--eds_button__icon__size, 24px)',
      width: 'var(--eds_button__icon__size, 24px)',
    },
  },
  states: {
    hover: {
      border: {
        type: 'border',
        width: '1px',
        color: 'transparent',
        radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
        style: 'solid',
      },
    },
    focus: {
      outline: {
        type: 'outline',
        offset: '2px',
        style: 'dashed',
        color: outline.color,
        width: outline.width,
      },
    },
    disabled: {
      background: disabledColor,
      border: {
        type: 'border',
        width: '1px',
        color: disabledBorderColor,
        style: 'solid',
      },
      typography: {
        ...buttonTypography,
        color: disabledTextColor,
        textAlign: 'center',
      },
    },
  },
  modes: {
    compact: {
      height: `var(--eds_button__height_compact, ${compactButtonHeight})`,
      spacings: {
        top: 'var(--eds_button__padding_y_compact, 0)',
        bottom: 'var(--eds_button__padding_y_compact, 0)',
      },
      clickbound: {
        height: compactClickboundHeight,
        width: '100%',
        offset: {
          top: `${
            (parseInt(compactClickboundHeight) -
              parseInt(compactButtonHeight)) /
              2 +
            1
          }px`,
          left: '0',
        },
      },
    },
  },
}
