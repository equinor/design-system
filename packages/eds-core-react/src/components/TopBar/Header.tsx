import { forwardRef, HTMLAttributes } from 'react'
import styled from 'styled-components'

export type TopbarHeaderProps = HTMLAttributes<HTMLDivElement>

const StyledHeader = styled.div`
  grid-area: left;
  display: grid;
  grid-template-columns: auto auto;
  grid-gap: 12px;
  align-items: center;
`

export const Header = forwardRef<HTMLDivElement, TopbarHeaderProps>(
  function Header({ children, ...props }, ref) {
    return (
      <StyledHeader ref={ref} {...props}>
        {children}
      </StyledHeader>
    )
  },
)

// Header.displayName = 'eds-topbar-header'
