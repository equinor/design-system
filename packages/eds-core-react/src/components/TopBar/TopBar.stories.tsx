import { useState, useEffect } from 'react'
import styled from 'styled-components'
import {
  TopBar,
  Icon,
  Search,
  TopbarProps,
  EdsProvider,
  Density,
  Typography,
} from '../..'
import { Story, ComponentMeta } from '@storybook/react'
import page from './TopBar.docs.mdx'

import {
  account_circle,
  accessible,
  notifications,
  fullscreen,
  apps,
} from '@equinor/eds-icons'

export default {
  title: 'Navigation/TopBar',
  component: TopBar,
  subcomponents: {
    Header: TopBar.Header,
    CustomContent: TopBar.CustomContent,
    Actions: TopBar.Actions,
  },
  parameters: {
    docs: {
      page,
    },
  },
} as ComponentMeta<typeof TopBar>

const icons = {
  account_circle,
  accessible,
  notifications,
  fullscreen,
  apps,
}

Icon.add(icons)

const Wrapper = styled.div.attrs({ tabIndex: 0 })`
  height: 65vh;
  overflow: auto;
`

const BodyWrapper = styled.div`
  height: 1000px;
  background: #ebebeb;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  padding: 16px;
`

export const Introduction: Story<TopbarProps> = (props): JSX.Element => {
  return (
    <Wrapper>
      <TopBar {...props}>
        <TopBar.Header>
          <>
            <Icon name="apps" />
            Application name - subtitle
          </>
        </TopBar.Header>
      </TopBar>
      <BodyWrapper>
        <Typography group="input" variant="text">
          Top of page
        </Typography>
        <Typography group="input" variant="text">
          Middle of page
        </Typography>
        <Typography group="input" variant="text">
          Bottom of page
        </Typography>
      </BodyWrapper>
    </Wrapper>
  )
}

export const WithSearchAndIcons: Story<TopbarProps> = (): JSX.Element => {
  const Icons = styled.div`
    display: flex;
    align-items: center;
    flex-direction: row-reverse;
    > * {
      margin-left: 40px;
    }
  `

  return (
    <TopBar>
      <TopBar.Header>
        <Icon name="apps" />
        Application name - subtitle
      </TopBar.Header>
      <TopBar.CustomContent>
        <Search aria-label="sitewide" id="search-normal" placeholder="Search" />
      </TopBar.CustomContent>
      <TopBar.Actions>
        <Icons>
          <Icon name="account_circle" size={16} title="user" />
          <Icon name="accessible" size={16} />
          <Icon name="notifications" size={16} />
          <Icon name="fullscreen" size={16} />
        </Icons>
      </TopBar.Actions>
    </TopBar>
  )
}
WithSearchAndIcons.storyName = 'With search and icons'

/* export const Compact: Story<TopbarProps> = (): JSX.Element => {
  const Icons = styled.div`
    display: flex;
    align-items: center;
    flex-direction: row-reverse;
    > * {
      margin-left: 40px;
    }
  `
  const [density, setDensity] = useState<Density>('comfortable')

  useEffect(() => {
    // Simulate user change
    setDensity('compact')
  }, [density])

  return (
    <EdsProvider density={density}>
      <TopBar>
        <TopBar.Header>
          <Icon name="apps" />
          Application name - subtitle
        </TopBar.Header>
        <TopBar.CustomContent>
          <Search
            aria-label="sitewide"
            id="search-normal"
            placeholder="Search"
          />
        </TopBar.CustomContent>
        <TopBar.Actions>
          <Icons>
            <Icon name="account_circle" size={16} title="user" />
            <Icon name="accessible" size={16} />
            <Icon name="notifications" size={16} />
            <Icon name="fullscreen" size={16} />
          </Icons>
        </TopBar.Actions>
      </TopBar>
    </EdsProvider>
  )
} */
