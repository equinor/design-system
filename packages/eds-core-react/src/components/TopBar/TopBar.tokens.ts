import { tokens } from '@equinor/eds-tokens'
import type { ComponentToken } from '@equinor/eds-tokens'

const {
  colors,
  spacings: { comfortable },
  typography: { navigation },
} = tokens

type TopBarToken = ComponentToken

export const topbar: TopBarToken = {
  height: '64px',
  background: colors.ui.background__default.rgba,
  typography: navigation.menu_title,
  spacings: {
    left: comfortable.large,
    right: comfortable.large,
    top: comfortable.small,
    bottom: comfortable.small,
  },
  border: {
    type: 'bordergroup',
    bottom: {
      style: 'solid',
      color: colors.ui.background__light.rgba,
      width: '2px',
    },
  },
  modes: {
    compact: {},
  },
}
