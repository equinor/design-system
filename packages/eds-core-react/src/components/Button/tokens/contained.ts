import { tokens } from '@equinor/eds-tokens'
import { ButtonToken } from '../Button.types'
import { button } from './button'

import { mergeDeepRight } from 'ramda'

const {
  colors: {
    text: {
      static_icons__primary_white: { rgba: primaryWhite },
    },
    interactive: {
      primary__resting: { rgba: primaryColor },
      primary__hover: { rgba: primaryHoverColor },
      secondary__resting: { rgba: secondaryColor },
      secondary__link_hover: { rgba: secondaryHoverColor },
      danger__resting: { rgba: dangerColor },
      danger__hover: { rgba: dangerHoverColor },
    },
  },

  shape: {
    button: { borderRadius: buttonBorderRadius },
  },
} = tokens

export const primary = mergeDeepRight(button, {
  background: primaryColor,
  typography: {
    color: primaryWhite,
  },
  border: {
    type: 'border',
    style: 'solid',
    width: '1px',
    color: primaryColor,
    radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
  },
  states: {
    hover: {
      background: primaryHoverColor,
    },
  },
}) as ButtonToken
export const secondary = mergeDeepRight(primary, {
  background: secondaryColor,
  border: {
    color: secondaryColor,
  },
  states: {
    hover: {
      background: secondaryHoverColor,
      border: {
        color: secondaryHoverColor,
      },
    },
  },
}) as ButtonToken

export const danger = mergeDeepRight(primary, {
  background: dangerColor,
  border: {
    color: dangerColor,
  },
  states: {
    hover: {
      background: dangerHoverColor,
      border: {
        color: dangerHoverColor,
      },
    },
  },
}) as ButtonToken
