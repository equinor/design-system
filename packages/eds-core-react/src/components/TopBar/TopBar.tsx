import { forwardRef, HTMLAttributes } from 'react'
import styled, { css, ThemeProvider } from 'styled-components'
import {
  spacingsTemplate,
  typographyTemplate,
  bordersTemplate,
  useToken,
} from '@equinor/eds-utils'
import { Paper } from '../Paper'
import type { Elevations } from '@equinor/eds-tokens'
import { topbar as topbarToken } from './TopBar.tokens'
import { useEds } from '../EdsProvider'

type AvailableElevations = keyof Pick<Elevations, 'none' | 'raised'>

export type TopbarProps = {
  elevation?: AvailableElevations
} & HTMLAttributes<HTMLDivElement>

const StyledTopBar = styled(Paper)<TopbarProps>(({ theme }) => {
  return css`
    height: ${theme.height};
    top: 0;
    position: sticky;
    background: ${theme.background};
    box-sizing: border-box;
    z-index: 1100;
    display: grid;
    grid-column-gap: ${theme.spacings.left};
    grid-template-columns: auto 1fr auto;
    grid-template-areas: 'left center right';
    align-items: center;
    ${bordersTemplate(theme.border)}
    ${spacingsTemplate(theme.spacings)};
    ${typographyTemplate(theme.typography)}
  `
})

export const TopBar = forwardRef<HTMLDivElement, TopbarProps>(function TopBar(
  { children, elevation = 'none', ...props },
  ref,
) {
  const { density } = useEds()
  const token = useToken({ density }, topbarToken)
  const rest = {
    ...props,
    ref,
  }
  return (
    <ThemeProvider theme={token}>
      <StyledTopBar forwardedAs={'header'} elevation={elevation} {...rest}>
        {children}
      </StyledTopBar>
    </ThemeProvider>
  )
})

// TopBar.displayName = 'eds-topbar'
