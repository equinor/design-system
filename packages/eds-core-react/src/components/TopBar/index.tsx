import { TopBar as BaseComponent, TopbarProps } from './TopBar'
import { Actions as TopbarActions, TopbarActionsProps } from './Actions'
import { Header as TopbarHeader, TopbarHeaderProps } from './Header'
import {
  CustomContent as TopbarCustomContent,
  TopbarCustomContentProps,
} from './CustomContent'

type TopbarCompoundProps = typeof BaseComponent & {
  Actions: typeof TopbarActions
  Header: typeof TopbarHeader
  CustomContent: typeof TopbarCustomContent
}

const TopBar = BaseComponent as TopbarCompoundProps

TopBar.Actions = TopbarActions
TopBar.Header = TopbarHeader
TopBar.CustomContent = TopbarCustomContent

TopBar.Actions.displayName = 'Topbar.Actions'
TopBar.Header.displayName = 'Topbar.Header'
TopBar.CustomContent.displayName = 'Topbar.CustomContent'

export { TopBar, TopbarActions, TopbarHeader, TopbarCustomContent }
export type {
  TopbarProps,
  TopbarActionsProps,
  TopbarCustomContentProps,
  TopbarHeaderProps,
}
