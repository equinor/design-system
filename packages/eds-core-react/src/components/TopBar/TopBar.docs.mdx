import { Links, PropsTable, Story } from './../../../.storybook/components'
import { TopBar } from '.'

# TopBar

Displays navigation and actions related to the interface.

<Links 
    ariaUrl="https://www.w3.org/WAI/ARIA/apg/patterns/toolbar/"
    figmaUrl="https://www.figma.com/file/0bGXR2sCwMVSDNyyzu5BXrO5/UI%E2%80%94User-Interface?node-id=17620%3A0"
    documentationUrl="https://eds.equinor.com/0b0c666ab/p/90b824-top-bar/b/0338b2"
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/TopBar/TopBar.tsx"
    npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react"
/>

<Story id="navigation-topbar--introduction" />
<PropsTable />

## Usage

Top bars span the full width of the viewport and are always the topmost component on a page.

<ul>
    <li>The top bar does not scroll away, being a consistent guidepost to users.</li>
    <li>The top bar always has a title which is the same throughout the whole application group.</li>
    <li>The top bar has one optional main navigational icon button which is always to the far left.</li>
</ul>

```tsx
import { TopBar } from '@equinor/eds-core-react'

<TopBar>
    <TopBar.Header>Application name</TopBar.Header>
</TopBar>
```

## Examples

### With search and icons

The actions on the far-right are standard.

<Story id="navigation-topbar--with-search-and-icons" />

<!--Re-enable when compact is implemented for TopBar-->
<!--### Compact

Compact `TopBar` using `EdsProvider`.

<Story id="navigation-topbar--compact" />-->
