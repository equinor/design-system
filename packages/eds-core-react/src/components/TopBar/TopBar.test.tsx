/* eslint-disable no-undef */
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'
import { axe } from 'jest-axe'
import styled from 'styled-components'
import { topbar as topbarToken } from './TopBar.tokens'
import { TopBar } from '.'

const { Actions, Header, CustomContent } = TopBar

const StyledTopBar = styled(TopBar)`
  background: red;
`

const ScrollContainer = styled.div`
  height: 2000px;
  width: 500px;
`
const { height } = topbarToken

describe('TopBar', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(
      <TopBar>
        <Header>
          <button type="button">Click me!</button>
        </Header>
        <CustomContent>
          <input type="text" />
        </CustomContent>
        <Actions>
          <p>icon</p>
        </Actions>
      </TopBar>,
    )

    expect(asFragment()).toMatchSnapshot()
  })
  it('Should pass a11y test', async () => {
    const { container } = render(
      <TopBar>
        <Header>Header tex</Header>
        <CustomContent>Content text</CustomContent>
        <Actions>
          <p>Icon</p>
        </Actions>
      </TopBar>,
    )
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Has correct height', () => {
    render(
      <TopBar data-testid="topbar">
        <Header>title</Header>
      </TopBar>,
    )
    const topbar = screen.getByTestId('topbar')

    expect(topbar).toHaveStyleRule('height', height)
  })

  it('Has all provided content', () => {
    const testIdHeader = 'topbar-test-header'
    const testIdCenter = 'topbar-test-center'
    const testIdActions = 'topbar-test-actions'

    render(
      <TopBar>
        <Header>
          <button type="button" data-testid={testIdHeader}>
            Click me!
          </button>
        </Header>
        <CustomContent>
          <input type="text" data-testid={testIdCenter} />
        </CustomContent>
        <Actions>
          <p data-testid={testIdActions}>icon</p>
        </Actions>
      </TopBar>,
    )
    expect(screen.getByTestId(testIdHeader)).toBeDefined()
    expect(screen.getByTestId(testIdCenter)).toBeDefined()
    expect(screen.getByTestId(testIdActions)).toBeDefined()
  })

  it('Has provided title', () => {
    const title = 'Some application title'
    render(<TopBar title={title} />)
    expect(screen.getByTitle(title)).toBeDefined()
  })

  it('Has sticky position in container', () => {
    const testId = 'topbar-test'
    render(
      <ScrollContainer>
        <TopBar className="test-bar" data-testid={testId}>
          Content
        </TopBar>
      </ScrollContainer>,
    )

    expect(screen.queryByTestId(testId)).toHaveStyleRule('position', 'sticky')
    expect(screen.queryByTestId(testId)).toHaveStyleRule('top', '0')
  })

  it('Can extend the css for the component', () => {
    render(<StyledTopBar data-testid="topbar"></StyledTopBar>)
    const topbar = screen.getByTestId('topbar')

    expect(topbar).toHaveStyleRule('background', 'red')
    expect(topbar).toHaveStyleRule('height', height)
  })
})
