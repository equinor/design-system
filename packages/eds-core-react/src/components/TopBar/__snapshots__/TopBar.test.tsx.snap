// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar Matches snapshot 1`] = `[Function]`;
